airbyte-integrations/connectors/source-woocommerce/unit_tests/integration/conftest.py

```diff
@@ -0,0 +1,134 @@
+    resource = stream_name if custom_resource is None else custom_resource
+
+    response_map = response
+    if response is None:
+        response_map = {"text": read_resource_file_contents(json_filename, __file__), "status_code": status_code}
+
+    return {
+        "request": {
+            "url": build_url(resource, modified_after=modified_after, modified_before=modified_before, is_regex=is_regex),
+            "is_regex": is_regex,
+            "method": method,
+        },
+        "response": response_map,
+    }
+
+
+def orders_http_calls():
+    return [request_response_mapping("orders", is_regex=True)]
+
+
+def coupons_http_calls():
+    return [request_response_mapping("coupons", is_regex=True)]
+
+
+def customers_http_calls():
+    return [request_response_mapping("customers")]
+
+
+def payment_gateways_http_calls():
+    return [request_response_mapping("payment_gateways")]
+
+
+def product_attributes_http_calls():
+    return [request_response_mapping("product_attributes", custom_resource="products/attributes")]
+
+
+def product_categories_http_calls():
+    return [request_response_mapping("product_categories", custom_resource="products/categories")]
+
+
+def product_reviews_http_calls():
+    return [request_response_mapping("product_reviews", custom_resource="products/reviews")]
+
+
+def products_http_calls():
+    return [request_response_mapping("products", is_regex=True)]
+
+
+def product_shipping_classes_http_calls():
+    return [request_response_mapping("product_shipping_classes", custom_resource="products/shipping_classes")]
+
+
+def product_tags_http_calls():
+    return [request_response_mapping("product_tags", custom_resource="products/tags")]
+
+
+def shipping_methods_http_calls():
+    return [request_response_mapping("shipping_methods", custom_resource="shipping_methods")]
+
+
+def shipping_zones_http_calls():
+    return [request_response_mapping("shipping_zones", custom_resource="shipping/zones")]
+
+
+def system_status_tools_http_calls():
+    return [request_response_mapping("system_status_tools", custom_resource="system_status/tools")]
+
+
+def order_notes_http_calls():
+    return [
+        request_response_mapping("orders", modified_after="2017-01-01.+", modified_before="2017-01-29.+", is_regex=True),
+        request_response_mapping("order_notes", custom_resource="orders/(727|723)/notes", is_regex=True),
+    ]
+
+
+def product_attribute_terms_http_calls():
+    return [
+        request_response_mapping("product_attributes", custom_resource="products/attributes", is_regex=True),
+        request_response_mapping("product_attribute_terms", custom_resource="products/attributes/.+/terms", is_regex=True),
+    ]
+
+
+def product_variations_http_calls():
+    return [
+        request_response_mapping("products"),
+        request_response_mapping("product_variations", custom_resource="products/(799|794)/variations", is_regex=True),
```

Comment:
naive question because I can't infer the context from the function: why do we expect the response to return either product with id 799 or 794? what's the deciding factor?

Reply:
This is one of the reason response builders are so powerful: they document the interactions between the different prerequisites (in this case, the interaction between the first response and the second request). Without those builders, the reader needs to assume that the values match something somewhere (where though?) which makes just reading the code of the test incomplete in order to fully comprehend what's going on. In other words, builders are more explicit.

They also allow for easier customization of scenarios while if I want a scenario that triggers pagination for example with the current setup, this is much harder to do

Reply:
Thanks, this is great feedback! I'll think how to address this kind of needs in future revisions, or how to integrate both approaches.